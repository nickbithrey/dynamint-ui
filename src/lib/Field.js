import React from 'react';
import { TextField } from 'office-ui-fabric-react/lib/TextField';
import { Checkbox } from 'office-ui-fabric-react/lib/Checkbox';

class Field extends React.Component {
	
	constructor(props) {
		super(props);
		this.state = {key: props.name, value: props.value};
		this.updateValue = this.updateValue.bind(this);
		this.update = this.update.bind(this);
	}
	
	updateValue(e) {
		e.preventDefault();
		this.setState({...this.state, value: e.target.value});
	}
	
	update(e) {
		this.props.update({[this.state.key]: this.state.value}, e);
	}
	
	render() {
		if (this.props.checkbox) {
			return <Checkbox {...this.props} defaultChecked={this.state.value} onChange={this.updateValue} onBlur={this.update} />
		}
		return (
			<TextField {...this.props} value={this.state.value} onChange={this.updateValue} onBlur={this.update} />
		);
	}
	
}

export default Field;