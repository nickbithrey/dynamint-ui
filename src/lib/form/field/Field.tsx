import * as React from 'react';
import { TextField } from './TextField';
import { CheckboxField } from './CheckboxField';
import { ListField } from './ListField';
import { IURILink } from '~/lib';

export const updateOnChange = <V extends IFieldValueType>(name: string, update: (name: string, value: V) => void) => (
    (e: React.ChangeEvent<HTMLInputElement>) => {
        e.preventDefault();
        update(name, e.target.value as unknown as V)
    }
);

export type IFieldType = 'text' | 'checkbox' | 'list';

export type IFieldValueType = string | number | boolean | Array<any> | IURILink;

export interface IField<V extends IFieldValueType> extends ITypedField<V> {
    type: IFieldType;
}

export interface ITypedField<V extends IFieldValueType> {
    name: string;
    label?: string;
    value: V;
    update?: (key: string, value: V) => void;
    entity?: any;
    [key: string]: any;
}

const fieldTypes: { [key in IFieldType]: React.ComponentType<any>} = {
    text: TextField,
    checkbox: CheckboxField,
    list: ListField,
};

export const Field = <V extends IFieldValueType>({type, ...remainingProps}: IField<V>) => {
    const FieldType: React.ComponentType<ITypedField<IFieldValueType>> = fieldTypes[type];
    return <FieldType {...remainingProps} />
}