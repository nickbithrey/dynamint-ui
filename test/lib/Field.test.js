import React from 'react';
import { shallow } from 'enzyme';

import Field from '~/lib/Field';

const Record = props => (
	<tr>
		<td>{props.reference}</td>
		<td>{props.description}</td>
	</tr>
);

describe('<Field /> rendering', () => {

	it('can render text field', () => {
		const props = {
			name: 'name',
			value: 'value',
			update: jest.fn(() => []),
			checkbox: false
		};
		const wrapper = shallow( <Field {...props} /> );
		
		expect(wrapper).toMatchSnapshot();
	});
	
	it('can render checkbox field', () => {
		const props = {
			name: 'name',
			value: 'value',
			update: jest.fn(() => []),
			checkbox: true
		};
		const wrapper = shallow( <Field {...props} /> );
		
		expect(wrapper).toMatchSnapshot();
	});
		
	it('renders no Checkbox when no checkbox parameter passed', () => {
		const props = {
			name: 'name',
			value: 'value',
			update: jest.fn(() => [])
		};
		const wrapper = shallow( <Field {...props} /> );
		
		expect(wrapper.find('StyledCheckboxBase')).toHaveLength(0);
		expect(wrapper.find('StyledTextFieldBase')).toHaveLength(1);
	});
		
	it('updates state when changed', () => {
		const props = {
			name: 'name',
			value: 'value',
			update: jest.fn(() => [])
		};
		const wrapper = shallow( <Field {...props} /> );
		
		expect(wrapper.find('StyledTextFieldBase')).toHaveLength(1);
		const event = {
			preventDefault: () => [],
			target: { value: 'newVal' }
		};
		wrapper.find('StyledTextFieldBase').simulate('change', event);
		expect(wrapper.instance().props.update).not.toBeCalled();
		expect(wrapper.state('value')).toBe('newVal');
	});
		
	it('calls update function on unfocus of element (on blur)', () => {
		const props = {
			name: 'name',
			value: 'value',
			update: jest.fn(() => [])
		};
		const wrapper = shallow( <Field {...props} /> );
		
		expect(wrapper.find('StyledTextFieldBase')).toHaveLength(1);
		expect(wrapper.find('StyledTextFieldBase').prop())
		const event = {
			preventDefault: () => []
		};
		wrapper.find('StyledTextFieldBase').simulate('blur', event);
		expect(wrapper.instance().props.update).toBeCalled();
		expect(wrapper.instance().props.update.mock.calls.length).toBe(1);
		expect(wrapper.instance().props.update.mock.calls[0][0]).toEqual({'name': 'value'});
		expect(wrapper.instance().props.update.mock.calls[0][1]).toBe(event);
	});
		
	it('calls update with current version of state', () => {
		const props = {
			name: 'name',
			value: 'value',
			update: jest.fn(() => [])
		};
		const wrapper = shallow( <Field {...props} /> );
		
		wrapper.setState({value: 'newVal'});
		expect(wrapper.find('StyledTextFieldBase')).toHaveLength(1);
		const event = {
			preventDefault: () => []
		};
		wrapper.find('StyledTextFieldBase').simulate('blur', event);
		expect(wrapper.instance().props.update).toBeCalled();
		expect(wrapper.instance().props.update.mock.calls.length).toBe(1);
		expect(wrapper.instance().props.update.mock.calls[0][0]).toEqual({'name': 'newVal'});
		expect(wrapper.instance().props.update.mock.calls[0][1]).toBe(event);
	});
		
});