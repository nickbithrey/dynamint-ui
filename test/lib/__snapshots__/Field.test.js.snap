// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Field /> rendering can render checkbox field 1`] = `
<StyledCheckboxBase
  checkbox={true}
  defaultChecked="value"
  name="name"
  onBlur={[Function]}
  onChange={[Function]}
  update={[MockFunction]}
  value="value"
/>
`;

exports[`<Field /> rendering can render text field 1`] = `
<StyledTextFieldBase
  checkbox={false}
  name="name"
  onBlur={[Function]}
  onChange={[Function]}
  update={[MockFunction]}
  value="value"
/>
`;
